### ISSUE TITLE
proposed 12 new codes for cheb/leg/vals/coeffs conversion

### ISSUE THREAD

**Issue by trefethen** [chebfun/chebfun#1580]
@nickhale and @ajt60gaibb have developed a remarkable collection of algorithms that speedily convert between four types of vectors of numbers:
```
    Chebcoeffs       Legcoeffs

    Chebvals         Legvals
```
There are 4x3=12 potential conversions here, of which only the Chebcoeffs<->Chebvals pair have well-established standard algorithms (namely the DCT).  Some of the 12 that might seem of questionable end-user value, like Legcoeffs -> Chebvals, turn out to be important steps in the algorithms for other conversions.

I propose that @nickhale and @ajt60gaibb implement twelve user-level Chebfun codes (or wrappers) with the following names:
```
chebcoeffs2chebvals   chebvals2chebcoeffs       "val/coeff conversions"
legcoeffs2legvals     legvals2legcoeffs

legcoeffs2chebcoeffs  chebcoeffs2legcoeffs      "leg/cheb conversions"
legvals2chebvals      chebvals2legvals

legcoeffs2chebvals    chebvals2legcoeffs        "cross conversions"
chebcoeffs2legvals    legvals2chebcoeffs
```
Many of these already exist in Chebfun at some level or another, but there isn't much uniformity in how they are packaged.  I think a uniform naming scheme is a good idea for the following reasons:

(1) Users will instantly know what names they can call.

(2) Everybody will suddenly find it easy to do all kinds of experiments.

(3) The uniform names will help organize our thinking, research, and publications related to these matters.

(4) These codes will immediately become benchmarks for the community.

I'm writing this on the assumption that Chebyshev points of the 2nd kind are the main issue, not points of the 1st kind.  If variants for 1st kind are desired this would bring the total number of potential conversions from 4x3=12 to 5x4=20, and I'm not sure I would go so far as to recommend that!  (If we do introduce some type1-related conversions, my recommendation would be to retain "chebvals" as meaning type 2 by default and introducing "cheb1vals" as the variant.)

**Comment by asgeirbirkis** [comment]
I agree that this would be an impressive functionality to have. One Chebfun design question is where those should live; as they won't take chebfun objects as arguments, they can't go into the `@chebfun` folder without becoming static (i.e. `chebfun.chebvals2legcoeffs()` ). I can think of three options:

1. Putting them as static methods in `@chebfun`.
2. Adding them into the `+cheb` package.
3. Adding a new package into Chebfun, i.e. `+fastTransforms`.

Out of those three option, the third has the appeal that it makes a nice clear package out of it. It'd also allow for a separate readme documentation inside the package about it.

**Comment by trefethen** [comment]
Isn't another option to put them in the root directory, like the current `leg2cheb`, `cheb2leg`, `chebsnake`, `scribble`, `cheblogo`, `chebpts` etc.?  It's nicer for users -- and more consistent with Matlab style conventions -- to be able to type `legvals2legpts` than `chebfun.legvals2legpts` or some such.  Moreover, these codes as I see it are "core Chebfun" in the sense that other things should be able to use them, right? (in contrast if I remember correctly to the convention with `+cheb`).

**Comment by asgeirbirkis** [comment]
Root directory is certainly an option, although wasn't there at some point discussion that we should avoid putting too many files there, so that users wouldn't open it up and be confused with a huge collection of files? Then again, this might be something that we want to draw users' attention to!

Also, what would happen if @ajt60gaibb and @nickhale were able to come up with fast transforms to other polynomials, hence increasing the number of boxes and connections in the diagram? Not sure if they are working on it, but if they obtain e.g. fast transforms to and from Chebyshev and Laguerre coeffs and vals, we'd have 6x5 = 30 files! 

I agree that other codes will want to use these methods (`conv` and `qr` are already applying algorithms from this school), so `+cheb` is not the correct place.

**Comment by aaustin141** [comment]
Closed in #1588.
